### PR TITLE
Improve local database

### LINKED ISSUE #22: Data download
Add option for data download for offline viewing
This may be automatic but in the prototype it is manual

- [ ] Button for download
- [ ] Data for that location will be downloaded and saved on local database

Possible:

- [ ] Toggle button for offline data download so only certai locations are automatic

### LINKED ISSUE #33: Finalise local database code
# Overview
The local database currently works, but is still in a prototype phase. Completing this issue will mean that the database is fully operational and ready to be used

## TODO
- [x] Refactor so that instead of saving ID number that is provided to use from metaweather, a new ID number is saved. 
Why do this? Because the provided ID number is pointless to us. The ID numbers are not consecutive when viewing entries and therefore they are displayed out of order in the recyclerView by default. To fix this we should create a brand new ID number for each new entry, that increases for each consecutive entry to keep the database ordered correctly
- [x] Once the ID numbers have been sorted out, the database now needs to save ALL the properties for each weather report (except the provided ID). Currently it is only saving the ID number (for prototyping purposes)
- [x] `SingleWeatherReportActivity` needs to be modified so that it can display entries from the database
- [x] Some form of **delete** functionality
- [x] Some form of **update** functionality (to be used if an update weather report is received for a day that has already been added to the database) - *after reviewing our app requirements, and the limitations of metaweather, this is probably not needed (as a feature of our app is the short term archival of weather forecasts - we probably want to keep all old reports rather than update with the current weather updates)*
- [x] Instead of displaying the ID number on the `recyclerView`, something more useful to the user should be displayed, such as the location name + date
- [x] Ensure documentation is up to date

### PR DESCRIPTION
**Pull request by JackMcGill**
# Overview 
This PR will merge code that improves upon our local database implementation. 

## Changes
- Automatically adds new weather reports to the local database, accounting for duplicate entries
- Deletes entries from the database when the delete button is tapped 
- Wraps weather reports retrieved from the database into a `Parcelable` object that can be added to an `Intent` so it can be passed to another activity. This logic has been previously explained in #10.
- Introduces an additional `id` property to the `WeatherReportModel` object. This change is explained in greater detail in the [database documentation](https://github.com/JackMcKill/CP3407-Project/blob/develop/Documentation/Database.md).

## Running this code yourself
As there have been changes to the structure of the database, you'll most likely need to delete any existing databases on your Emulator before you can run this code successfully - running this code on an Emulator with an older version of the database will cause an error saying  "*Room cannot verify the data integrity. Looks like you've changed schema but forgot to update the version number. You can simply fix this by increasing the version number.*".  

### To delete any exisiting databases
1. Open the `Device File Explorer` (found in the bottom right side of Android Studio) 
Screen Shot 2021-10-09 at 4 15 09 pm
2. Open `Data` `->` `Data` `->` `com.cp3407.wildernessweather` `->` `databases`
3. Delete the contents of the `databases` folder
Screen Shot 2021-10-09 at 4 18 58 pm
4. Restart the emulator, and you should be good to go!

### To check the actual functionality
Compared the last iteration of the database, it probably *looks* same at first glance, but the way entries are added are very different. 
1. After you have cleared your previous database, you can run the app and check the contents of the database by tapping the `Press for DB Prototype` button. This should show a blank screen (because the database is empty). 
2. Then you can view a weather report for any location and day using the `APIActivity` button. Once you are viewing the `SingleWeatherReportActivity` screen, that weather report has been added to the database automatically. 
3. You can check by revisiting the `Press for DB Prototype`. You should see a new entry. 
4. Now you can check whether the code automatically checks for exisiting entries before adding another one - just search for the same location and click on the same weather report to bring back the `SingleWeatherReportActivity`. Now you can go back once more to the database screen and if everything worked correctly, there should be **no** new entries to the database. 
5. Finally, you can check the delete functionality is working correctly by pressing on the `X` on a database item to delete it from the database.

In my testing, I found that [metaweather.com](https://www.metaweather.com) updates the weather information roughly every 6 hours. This means that if you were to keep searching for the weather in Brisbane, and viewing the first weather report, it will be the exact same report for 6 hours straight (and therefore will not be added to the database). However, at some point after roughly 6 hours, a new report will be posted by [metaweather.com](https://www.metaweather.com), and if you were still repeatedly viewing that same result in our app, it's at this point that a new entry will be added to the database.  
This is because each new weather report gets it's own unique `id` number (provided by [metaweather.com](https://www.metaweather.com)), and whether or not a database entry is considered a duplicate or not is based off a comparison of the current weather report being viewed's `id` number, and the `id` number of all entries already in the database.

*Did that make* any *sense at all?* If not, feel free to ask me, as I can probably explain it better over the phone or in a call than over text like this 😆.

## Notes

Something that I forgot to add into this PR was to change the clickable area of the items displayed in the `RecyclerView`. Currently, only the `id` number on the left of each item is clickable. Tapping on the rest of the item does nothing.
Screen Shot 2021-10-09 at 4 20 28 pm

This is a small fix, I will make it into an issue to be solved later - **SOLVED in dd7319f**

Merging this PR will close #33 & close #22.

### CODE REVIEW

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (HarperThurlow) [inline app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:22]: I think our coding practice was to have "tv_cityName" instead of "cityNameView". I know this is the same as a the current xml file but I think this was the same.

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (HarperThurlow) [inline app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:157]: Is this meant to be blank?

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (HarperThurlow) [inline app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:164]: Function for later?

**Review by HarperThurlow — APPROVED** [review]
From GitHub I have written a few comments on the files. I cannot currently run the code in AndroidStudio unfortunately but from my perspective the code looks good. If this isn't stopping you from continuing on with different work, please leave this PR open and don't merge into develop just yet so I can test (hopefully soon) in AndroidStudio :)

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (JackMcGill) [reply app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:164]: Same as above - this is Calebs code.
Currently blank, will be worked on later, probably by @CalebWebsterJCU

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (JackMcGill) [reply app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:157]: This is Calebs code from #30. It's appearing here in my Pull request because I rebased my branch onto his so I could have his changes to work with, before he merged his branch.

Currently blank, will be worked on later, probably by @CalebWebsterJCU

File: app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java
Comment (JackMcGill) [reply app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:22]: Same as previous comments - this is Calebs code

In `Contributing` we state that they have the type-prefix when declared in the XML code, so they are easily identified when using `findViewByID` in the Java code - not specifically in the Java code (where camel case is the codestyle).

I _personally_ prefer it this way - having the type-prefix in the XML only so they can be identified quickly inside the Java code with `findViewById`, and then the Java variables named to something suitable for their purpose, but I'm  open to enforcing the type-prefix everywhere. What do you think?

Comment (JackMcGill) [reply app/src/main/java/com/cp3407/wildernessweather/SingleWeatherReportActivity.java:22]: It sounds kinda contradictory when I write it out like that lol. Happy to open this up into a discussion

**Comment by JackMcGill** [comment]
> please leave this PR open and don't merge into develop just yet so I can test 

Of course, PR will be remaining open until you have a chance to run it yourself. Once you've run it I'll resolve the conflicts (if I resolve them now, they will likely just need to be re-resolved again at a later date after new code has been pushed to develop).

**Pull request merged**
